src/collapse/perf.rs

```diff
@@ -324,6 +354,87 @@ impl PerfState {
         }
         Ok(())
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> Result<(), String> {
+        let addr = u64::from_str_radix(addr, 16)
+            .map_err(|_| format!("unable to parse {} as hex address", addr))?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let file =
+                File::open(module).map_err(|_| format!("unable to open file: {}", module))?;
+            // Using memmap is unsafe because the underlying file may be modified,
+            // which would cause undefined behavior. It's unlikely to be a problem in practice
+            // since we shouldn't run very long. The potential speed-up is worth it.
+            let map = unsafe {
+                memmap::Mmap::map(&file)
+                    .map_err(|_| format!("unable to memmap file: {}", module))?
+            };
+            let file = &object::File::parse(&*map)
+                .map_err(|_| format!("unable to parse file: {}", module))?;
+            entry.insert(addr2line::Context::new(file).map_err(|e| e.to_string())?);
+        };
+        let ctx = &self.addr2line_contexts[module];
```

Comment:
It's a little unfortunate that we can't save the extra map lookup here due to the need to return early on error. If it weren't for that, we could just use `.entry().or_insert_with(|| {})`. I wonder if there's a way to get around that. It's like we want a `fallible_or_insert_with` :/ Even moving to a setup where we _don't_ return the error but just print it won't help, since we'll still have to return early.

Reply:
Actually, here's a though: what if we make the map `HashMap<String, Option<Context>>`? That way, we can do:
```rust
let ctx = self.addr2line_contexts.entry(module.into()).or_insert_with(|| {
    let file = File::open(module).map_err(|e| {
        warn!("unable to open module file {} to resolve {}: {:?}", module, addr, e);
    }).ok()?;
    // ...
    addr2line::Context::new(file).map_err(|e| {
        warn!("could not parse debug symbols from module {}: {:?}", module, e);
    }).ok()
});
let ctx = if let Some(ctx) = ctx {
    ctx
} else {
    return;
};
```
This has the added advantage that we won't keep trying to resolve debug symbols for a module once we've tried to parse it once!
This

Reply:
Yes, that's great! I really wanted to use or_insert_with, but couldn't find a nice way to make it work.